contxt/auth/__init__.py

```diff
@@ -1,142 +1,107 @@
-        token = self.contxt_auth.get_new_token_for_audience(service_audience)
-        self.store_service_token(audience=service_audience,
-                                 access_token=token)
-        return token
-
-    def get_token_for_audience(self, audience):
-        logger.debug(f"Getting token for client {audience}")
-        # check to see if we've gotten a token for this service or not
-        if audience not in self.tokens:
-            # try to get a token, whether it's a refresh or not
-            self.authenticate_to_service(audience)
-
-        # check to see if have the token, but needs to be refreshed
-        if self.token_is_expired_for_audience(audience):
-            logger.warn(f"Refreshing expired token for client {audience}...")
-
-            # if it's the contxt auth client, we need to follow the other refresh route via Auth0
-            if audience == AUTH_AUDIENCE:
-                self.refresh_contxt_auth_token()
-            else:
-                self.authenticate_to_service(audience)
-
-        access_token = self.tokens[audience]['token']
-        return access_token
-
-    def token_is_expired_for_audience(self, audience):
-        logger.debug(f"Checking if token is expired for audience {audience}")
-        if not self.tokens.get(audience):
-            return True
-        access_token = self.tokens[audience]['token']
-        decoded_token = decode(access_token, verify=False)
-        token_expiration_epoch = decoded_token['exp']
-
-        return token_expiration_epoch <= Utils.get_epoch_time(datetime.now())
-
-    def read_token_file(self):
-        logger.debug(f"Loading tokens from {self.token_file}")
-        self.contxt_config_path.mkdir(parents=True, exist_ok=True)
-
-        try:
-            with self.token_file.open('r') as f:
-                return load(f)
-        except FileNotFoundError:
-            logger.debug("Token file does not yet exist")
-            return {}
-
-    def load_tokens(self):
-        all_tokens = self.read_token_file()
-        return all_tokens.get(self.client_id, {})
-
-    def store_service_token(self, audience, access_token, refresh_token=None):
-
-        self.tokens.setdefault(self.client_id, {})
-        # if self.client_id not in self.tokens:
-        #     self.tokens[self.client_id] = {}
-
-        self.tokens[audience] = {
-            'token': access_token,
-            'refresh_token': refresh_token
-        }
-
-        self.store_tokens(self.tokens)
-
-    def store_tokens(self, tokens_for_client):
-        logger.debug(f"Storing tokens to {self.token_file}")
-
-        all_tokens = self.read_token_file()
-        all_tokens[self.client_id] = tokens_for_client
-
-        with self.token_file.open("w") as f:
-            dump(all_tokens, f, indent=4)
-
-    def clear_tokens(self):
-        logger.debug(f"Removing toke file {self.token_file}")
-        self.token_file.unlink()
+
+    def get_token_provider(self, audience: str) -> TokenProvider:
```

Comment:
this method could be abstract, this implementation seems to be the machine based one, right?

Reply:
right, i'll change that